index.js

```diff
@@ -678,41 +681,22 @@
   };
 
   //  commonTypes :: ([[Type]], Boolean) -> [Type]
-  var commonTypes = function(typeses, loose) {
-    var types = chain(typeses, id);
-    if (isEmpty(types)) return [];
-
-    var $1s, $2s, candidates, idx, t;
-    if (types[0].type === 'UNARY') {
-      $1s = {};
-      for (idx = 0; idx < types.length; idx += 1) {
-        t = types[idx];
-        if (t.$1.type !== 'UNKNOWN') $1s[t.$1.name] = t.$1;
-      }
-      candidates =
-      map(or(map(keys($1s), function(k) { return $1s[k]; }), [Unknown]),
-          UnaryType.from(types[0]));
-    } else if (types[0].type === 'BINARY') {
-      $1s = {};
-      $2s = {};
-      for (idx = 0; idx < types.length; idx += 1) {
-        t = types[idx];
-        if (t.$1.type !== 'UNKNOWN') $1s[t.$1.name] = t.$1;
-        if (t.$2.type !== 'UNKNOWN') $2s[t.$2.name] = t.$2;
-      }
-      candidates = BinaryType.xprod(
-        types[0],
-        or(map(keys($1s), function(k) { return $1s[k]; }), [Unknown]),
-        or(map(keys($2s), function(k) { return $2s[k]; }), [Unknown])
-      );
-    } else {
-      candidates = types;
-    }
-
-    return filter(candidates, function(t1) {
-      return all(typeses, function(types) {
-        return any(types, function(t2) {
-          return equalTypes(t1, t2, loose);
+  //
+  //  [[String, RegexFlags], [String, RegexFlags]] ---> [String, RegexFlags]
+  //
+  //  [[Boolean], [Boolean], [Boolean], [Number]] ---> []
+  //
+  //  [[Array ???], [Array String], [Array ???]] ---> [Array String]
+  //
+  //  [[Either String ???], [Either ??? Number]] ---> [Either String Number]
+  var commonTypes = function($typeses, loose) {
+    return filter($typeses[0], function($t) {
+      return all($typeses, function(types) {
+        return any(types, function(t) {
+          var equal = equalTypes($t, t, loose);
+          if (equal && '$1' in $t && $t.$1.type === 'UNKNOWN') $t.$1 = t.$1;
+          if (equal && '$2' in $t && $t.$2.type === 'UNKNOWN') $t.$2 = t.$2;
```

Comment:
Currently it's not only a mutation, it's a mutation of something shared (the input object). The caller seems to be aware of this (since it also prefixes the variable with `$`, but doesn't seem to benefit from, or use the mutation in any way. (Correct me if I'm wrong)

Wouldn't it be better to contain the mutation within the scope of `commonTypes`? That way The easiest way to do this would be deep-cloning the `$typeses[0]` object before you start filtering it. Alternatively, I think you can stay away from mutation altogether by using `reduce` for your iterations. If you want I can write the required reducers

---

I think you can do `tail` on `$typeses` before passing it to `all`. It would save you a single iteration, or in cases of a unary array, all of the work.

---

By using `any` to iterate the types, you're not actually collecting all the type info you *could*. For example if you have `[[Left String], [Left String, Right Number]]`, I think the output would be `[Either String ???]`(correct me if I'm wrong)  because `any` kept short-circuiting before you reached the `Right`. This too could be resolved with some custom `reduce`, not sure if you'd want to though.